wp-content/plugins/wporg-learn/js/hooks.js

```diff
@@ -0,0 +1,27 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+
+export const useIsBlockCompatibleWithSidebar = ( attributes, sidebarName ) => {
+	let widgets = {};
+
+	useSelect( ( select ) => {
+		widgets = select( 'core/edit-widgets' ).getWidgets();
+	} );
+
+	const widgetId = attributes.__internalWidgetId;
+
+	// __internalWidgetId is not stable.
+	// If it is undefined we can't look up the widget sidebar type to check compatibility
+	// so we have to return true and let the backend validation handle the incompatible type.
+	if ( widgetId === undefined ) {
+		return true;
+	}
```

Comment:
@ryelle any opinion or suggestions for improving this? I'm guarding against it being removed but I'm not sure whether I should be referencing it at all and I haven't found any other way of linking the current block to the list of widgets so far...

Reply:
I don't think this is working correctly, I can add the block from any widget area & while I saw the warning once, it doesn't seem consistent.

I pushed up a commit that changes this to check the block ancestry, since technically each widget area is a `core/widget-area` block, and the `id` attribute of that block is the sidebar name. I also changed the hook name to `useIsBlockInSidebar`, to reflect better what it's doing. This seems to work for widgets and the regular post editor, even if the block is nested in other blocks (group, etc).

If I misunderstood the purpose of this function, feel free to drop my commit :)

Reply:
Code wise this looks great thanks, lots for me to learn about these APIs!

I'll test it out...

Reply:
> This seems to work for widgets and the regular post editor,

I assumed incorrectly that a block with `category: 'widgets'` would only be allowed in the widget editor, it's a shame that's not the case. Are you aware of a programmatic way of blocking the post editor, or is my approach of using a message all that should be expected?

Reply:
No, categories are just categories — you can see the "widgets" category in the post editor too. Looks like [the `ancestor` property](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#ancestor) in `block.json` could work, if you add `ancestor: [ "core/widget-area" ]` it _should_ restrict it to the widget editor.

Reply:
Decided there's really no reason why this info couldn't be displayed in a post, so it won't be blocked, but I've just added a new commit to show different messages depending on where it's used.

Reply:
| Lesson Plan Post | Workshop Post | Lesson Plan Sidebar | Workshop Sidebar |
|-|-|-|
| ![Screen Shot 2022-09-28 at 5 30 12 PM](https://user-images.githubusercontent.com/1017872/192689143-eddce84a-37df-40fc-8a6f-1b9c7180d142.jpg) | ![Screen Shot 2022-09-28 at 5 30 28 PM](https://user-images.githubusercontent.com/1017872/192689165-00ea7f11-9cb8-4019-8138-05555d2ff91d.jpg) | ![Screen Shot 2022-09-28 at 5 30 48 PM](https://user-images.githubusercontent.com/1017872/192689189-51927a02-a7b7-4859-a643-fac0893215a3.jpg) | ![Screen Shot 2022-09-28 at 5 31 04 PM](https://user-images.githubusercontent.com/1017872/192689215-a1ff6dec-ded2-4b69-878b-be45fe757bed.jpg) |